Develop

msiarko opened the pull request:
Features:
- Register/Login users;
- Add, remove, edit book list items;
- Edit book list;
- Reding time tracking.

msiarko committed:
Login page and API feature is implemented;
Register page is implemented.

msiarko committed:
Code refactoring

msiarko committed:
Unneseccary files is deleted

msiarko committed:
Code refactoring

msiarko committed:
Role claim is added

msiarko committed:
Updated project structure

msiarko committed:
API:
- added base CQRS functionality;
- performed code refactoring

Client:
- enabled CSS modules;
- added ContextMenu component;
- refactored project structure.

msiarko committed:
Added unversioned files

msiarko committed:
Removed less files

msiarko committed:
Perfomed code refactoring

Commit:
Updates was performed:
- Added edit mode to list item;
- Code refactoring was performed.

Commit:
Updates performed:
- Added edititng, saving and cancel functioanality;
- Refactored ContextMenu component;
 - request reducer and actions are replaced by  loading reducer and actions.

Commit:
Code refactoring

Commit:
Updates was performed:
- Added read and write model projects;
- ReadingList.Client Code refactoring.

msiarko committed:
Updated package.json file

msiarko committed:
Updates performed:
- Added data acces infrastructure;
- Code refactoring.

Commit:
Updates performed:
- Code refactoring;
- Initial migration added.

msiarko committed:
Code refactoring

msiarko committed:
Updates performed:
- Code refactoring;
- Name property added to BookList entity

Commit:
Updates performed:
- Code refactoring;
- Tests added.

msiarko committed:
Code refactoring is performed and PrivateList query implementation is added

Commit:
Updates performed:
- Private list operations implemented on client and server sides;
- Code refactoring.

Commit:
Updates performed:
- Code refactoring;
- Private list items CRUD operations is implemented.

Commit:
Updates performed:
- Code refactoring.

Commit:
Updates performed:
- Code refactoring.

Commit:
Updates performed:
- Code refactoring.

Commit:
Updates performed:
- Code refactoring;
- Private list name edit implemented.

Commit:
Code refactoring

Commit:
Updates performed:
- Code refactoring;
- Loading spinner added.

Commit:
Updates performed:
- Code refactoring;
- Resources added;
- Custom exceptions added;
- UserSqlService added.

Commit:
Updates performed:
- Code refactoring.

Commit:
Updates performed:
- Code refactoring.

Commit:
Updates performed:
- Update entity extension added.

msiarko committed:
Name of RedButton.css.d.ts was fixed

Commit:
Updates performed:
- Code refactoring;
- ReadingList.Client.njsproj added.

Commit:
Updates performed:
- Unnecessary files removed

Commit:
ReadingJournalRecord added

Commit:
Implementation of reading time calculation started

msiarko committed:
Updates performed:
- Code refactoring;
- Reading time tracking added.

msiarko committed:
Updates performed:
- Code refactoring;
- Unit tests added.

Commit:
Code refactoring is performed

Commit:
Code refactoring is performed

Commit:
Code refactoring is performed

KatzuoOgust commented on ReadingList.Domain/DTO/BookList/SharedBookListDto.cs:
```diff
@@ -0,0 +1,23 @@
+    {
+        [JsonIgnore]
+        public int Id { get; set; }
+        [JsonIgnore]
```
add new line between props

KatzuoOgust commented on ReadingList.Domain/MapperProfiles/BookListProfile.cs:
```diff
@@ -0,0 +1,23 @@
+                    JsonSerializer.Create(new JsonSerializerSettings
+                    {
+                        Formatting = Formatting.Indented
+                    })))).ReverseMap().ConvertUsing<BookListConverter>();
```
chop line by dots as hard to see what is for what:
```cs
CreateMap<SharedBookListDto, BookList>()
	.ForMember(d => d.JsonFields,
		m => m.MapFrom(
			b => JObject.FromObject(b, JsonSerializer.Create(new JsonSerializerSettings
				{
					Formatting = Formatting.Indented
				}))))
	.ReverseMap()
	.ConvertUsing<BookListConverter>();
```

KatzuoOgust commented on ReadingList.Domain/Services/Encryption/DefaultEncryptionAlgorithm.cs:
```diff
@@ -0,0 +1,32 @@
+            };
+        }
+
+        public string Execute(string normalString) => _algorithm(normalString);
```
why not to move lambda body as method body to that method?

KatzuoOgust commented on ReadingList.Domain/Services/ReadingTimeCalculator.cs:
```diff
@@ -0,0 +1,30 @@
+            {
+                case BookItemStatus.StartedButPostponed:
+                case BookItemStatus.Read:
+                    diff = TimeSpan.FromTicks(DateTime.Now.Ticks - previousDate.Ticks);
```
No need to get ticks, just do: `diff = DateTime.Now - previousDate`

KatzuoOgust commented on ReadingList.Domain/Services/Sql/Implementations/UserSqlService.cs:
```diff
@@ -0,0 +1,27 @@
+                .From("Users");
+        }
+
+        public string GetUserByLoginSql(string parameterName = "login") =>
```
not sure that `parameterName` is a right name for argument  
can be named as login
and what was the intention to defined default value `login`?

KatzuoOgust commented on ReadingList.Domain/Services/Sql/Implementations/UserSqlService.cs:
```diff
@@ -0,0 +1,27 @@
+        }
+
+        public string GetUserByLoginSql(string parameterName = "login") =>
+            _sqlBuilder.Where($"Login = @{parameterName}").ToSql();
```
potential place for SQL injection

KatzuoOgust commented on ReadingList.Api/Controllers/AccountController.cs:
```diff
@@ -0,0 +1,49 @@
+            
+            var result = await AskAsync(new LoginUserQuery(loginData.Email, loginData.Password));
+
+            if (!result.IsSucceed)
```
Code was repeated several times so can be moved into separate method so will look like (just an example):
```cs
return AskResult(result);
```

Commit:
Updates performed:
- Removed Result and Result<T> classes;
- Implemented ExceptionHandlingMiddleware;
- Implemented mapping exception type to status code;
- Code refactoring;
- Fixed some issues found by Sergey Svirsky.

Commit:
Updates performed:
- Registered ExceptionToStatusCodeProvider as singleton in DI;
- Refactored ExceptionHandlingMiddleware to use ExceptionToStatusCodeProvider as injected service;
- Replaced ObjectNotFoundException with ObjectNotExistException (the ObjectNotFoundException will be used only for search queries).

KatzuoOgust commented on ReadingList.Api/ExceptionToStatusCodeProvider.cs:
```diff
@@ -0,0 +1,51 @@
+                if (_source.TryGetValue(exceptionType, out var statusCode))
+                    return statusCode;
+                exceptionType = exceptionType.BaseType;
+            }
```
`while(true)` лучше юзать когда ты реально НЕ знаешь основное условие выхода
тут же можно юзать `exceptionType.IsSubclassOf(_baseCaseExceptionType)`
читая же код с `while(true)` начинаешь думать что цикл может быть бесконечным когда это не так
```cs
while(exceptionType.IsSubclassOf(_baseCaseExceptionType)) 
{

}
return HttpStatusCode.InternalServerError;
```

KatzuoOgust commented on ReadingList.Api/ExceptionToStatusCodeProvider.cs:
```diff
@@ -0,0 +1,51 @@
+        {
+            return
+                map.SelectMany(pair => pair.Value.Select(t => new KeyValuePair<Type, HttpStatusCode>(t, pair.Key)),
+                    (p, c) => c).Distinct(new KeyValueEqualityComparer()).ToDictionary(pair => pair.Key, pair => pair.Value);
```
если вдруг появляется перенос аргумента на новую строку то и цепочку вызовов лучше разносить на новые строки для повышения читабельности иначе с ходу не понятно кто за кем чейнится и чейнимся ли мы на результат или какой-то аргумент

KatzuoOgust commented on ReadingList.Api/ExceptionToStatusCodeProvider.cs:
```diff
@@ -0,0 +1,51 @@
+        public ExceptionToStatusCodeProvider(Dictionary<HttpStatusCode, Type[]> map)
+        {
+            _baseCaseExceptionType = typeof(Exception);
+            _source = InitializeExceptionToStatusCodeMap(map);
```
т.к. внутри метода инициализациия содается новый словарь то нет смысла делать аргумент столь конкретным и можно указать, например, тип IReadOnlyDictionary

KatzuoOgust commented on ReadingList.Api/Middlewares/ExceptionHandlingMiddleware.cs:
```diff
@@ -0,0 +1,40 @@
+            _next = next;
+        }
+
+        public async Task InvokeAsync(HttpContext context, ExceptionToStatusCodeProvider exceptionToStatusCodeProvider)
```
`ExceptionToStatusCodeProvider ` в данном случае лучше будет передавать не в `InvokeAsync` метод а в конструктор, т.е. в момент `app.UseMiddleware`

KatzuoOgust commented on ReadingList.Api/Middlewares/ExceptionHandlingMiddleware.cs:
```diff
@@ -0,0 +1,40 @@
+            {
+                await _next(context);
+            }
+            catch (Exception e)
```
ловить все Exception-ы не очень ибо в случае тасков и асинк кода
при канселейшне может быть выкинут `OperationCanceledException` ([link](https://msdn.microsoft.com/ru-ru/library/system.threading.cancellationtoken.throwifcancellationrequested(v=vs.110).aspx)) на который скорее всего ответить не сможешь либо нет смысла

KatzuoOgust commented on ReadingList.Client/src/components/NavBar/NavBar.tsx:
```diff
@@ -0,0 +1,26 @@
+}
+
+const NavBar: React.SFC<NavBarProps> = props => {
+    const navLinks = props.links.map((value, index) => {
```
а можно return в лямбде опустить и записать чуть проще

KatzuoOgust commented on ReadingList.Client/src/components/NavBar/NavBarLink/NavBarLink.tsx:
```diff
@@ -0,0 +1,28 @@
+
+const NavBarLink: React.SFC<NavBarLinkProps> = props => {
+    const link = props.link.action
+        ? (<a className={props.className} onClick={props.link.action}>{props.link.text}</a>)
```
большая тернарка: либо меняем тернарку на if else либо делаем частные методы рендеринга
`const link = props.link.action ? renderA() : renderB();`

KatzuoOgust commented on ReadingList.Client/src/components/NavBar/NavBarLink/NavBarLink.tsx:
```diff
@@ -0,0 +1,28 @@
+        ? (<a className={props.className} onClick={props.link.action}>{props.link.text}</a>)
+        : (
+            <NavLink className={props.className} activeClassName={props.activeClassName}
+                        to={props.link.href}>
```
в ряде реакт гайдов можно встретить совет касающийся стилистики кода:
 если вдруг мы передаем какие-то пропы и вынуждены переносить то лучше переносить каждый проп
```jsx
<NavLink
  className={props.className}
  activeClassName={props.activeClassName}
  to={props.link.href}
>
  {content}
</NavLink>
```

KatzuoOgust commented on ReadingList.Client/src/components/PrivateBookLI/PrivateBookLI.tsx:
```diff
@@ -0,0 +1,73 @@
+        return (
+            <li className={styles['editing-book-li']} >
+                <form onSubmit={props.onChangesSubmit}>
+                    <input type="hidden" name="item-id" value={props.listItem.id} />
```
для большей наглядности можно делать при необходимости локальные Footer,Header,TitleEdit и прочие компоненты либо юзать renderSmth методы

KatzuoOgust commented on ReadingList.Client/src/containers/Account/Account.tsx:
```diff
@@ -0,0 +1,119 @@
+        const email = (form.elements.namedItem('email') as HTMLInputElement).value;
+        const password = (form.elements.namedItem('password') as HTMLInputElement).value;
+        const confirmPassword = form.elements.namedItem('confirmPassword') as HTMLInputElement;
+        if(confirmPassword === null) {
```
эта логика могла бы быть внутри `AccountForm` где она и должна быть.
сейчас получается что `Account` знает про некоторое устройство иных компонент.
onSubmit имел бы примерно такую сигнатуру:
`({ emaill: string, password: string, confirmPassword: bool }) : void`

KatzuoOgust commented on ReadingList.Client/src/containers/PrivateBookList/PrivateBookList.tsx:
```diff
@@ -0,0 +1,195 @@
+        return (
+            <div>
+                {
+                    this.props.bookList && !this.props.bookList.isInEditMode 
```
в JSX сложные условия лучше выносить в `const` / `var` переменные либо можно даже в геттеры для читабельности.
конкретный случай можно оставить как есть

msiarko committed:
Updates performed:
- Updated connection string;
- Added filtering of OperationCanceledException in ExceptionHandlingModdleware;
- Moved ExceptionToStatusCodeProvider from InvokeAsync method to constructor;
- Updated GetStatusCode method in ExceptionToStatusCodeProvider;
- Code refactoring of Client project.

Commit:
Updates performed:
- PrivateBookList.tsx was code refactored;
- Added filtering of responses with code status 401;
- Added PrivateListNameEditor component.

Commit:
Updates performed:
- ExceptionHandlingMiddleware was code refactored;
- Fixed not authenticated message;
- Added list of browsers where context menu issue had occurred.

Commit:
Updates performed:
- AccountController.cs was code refactored;
- withContextMenu HOC was refactored for fixing issue with wrong calling area of context menu.

Commit:
Updates performed:
- Fixed  issue with displaying multiple context menues at the same time.

Commit:
Updates performed:
 - Code refactoring;
 - Fixed potential issue with single instance of SqlBuilder in UserSqlServer.

msiarko committed:
Updates performed:
 - Added PrivateBookListSqlService and EntitiesValidator;
 - Code refactoring.

msiarko committed:
Updates performed:
 - Moved TryGetNotExistExceptionMessageByTypeName method from extension of ResourceManager to EntitiesValidator class as private static member;
 - Code refactoring.

msiarko committed:
Updates performed:
- Code refactoring.

msiarko committed:
Updates performed:
 - Added checkings of arguments in some functions;
 - Code refactoring.

KatzuoOgust commented on ReadingList.Api/Abstractions/BaseController.cs:
```diff
@@ -0,0 +1,51 @@
+
+        protected BaseController()
+        {
+            _mediator = MediatorContainer.GetMediator();
```
почему решил использовать servicelocator подход для получения инстанса медиатора вместо DI через конструктор?

msiarko replied:
чтобы не создавать одинаковые конструкторы в классах наследниках, ибо по замыслу контроллеры будут использовать только медиатор для взаимодействия с доменной логикой

KatzuoOgust commented on ReadingList.Api/Controllers/AccountController.cs:
```diff
@@ -0,0 +1,39 @@
+    {
+        [HttpPost]
+        [Route("login")]
+        [ValidateModelState]
```
учитывая что эта проверка навешена на каждый экшон контроллера то можно мувнуть на уровень контроллера(т.е. класса)

KatzuoOgust commented on ReadingList.Api/Controllers/PrivateListController.cs:
```diff
@@ -0,0 +1,67 @@
+        
+        [HttpPut("items/{id}")]
+        [ValidateModelState]
+        public async Task<IActionResult> Put([FromRoute] int id, [FromBody] UpdatePrivateListItemData updatePrivateListItemData)
```
а не лучше ли разграничить два метода Put по названию а не по аргументам
один оставив как Put (продразумеватся Put Private List) 
а второй превратив в UpdateItem или поднобным образом?

KatzuoOgust commented on ReadingList.Api/ExceptionToStatusCodeProvider.cs:
```diff
@@ -0,0 +1,56 @@
+        private static IReadOnlyDictionary<Type, HttpStatusCode> InitializeExceptionToStatusCodeMap(IReadOnlyDictionary<HttpStatusCode, Type[]> map)
+        {
+            return
+                map.SelectMany(pair => pair.Value
```
а почему используешь  `SelectMany` с collection & result селекторами если вроде как достаточно использовать с просто селектором
```cs
  map.SelectMany(pair => pair.Value.Select(t => new KeyValuePair<Type, HttpStatusCode>(t, pair.Key)))
                    .Distinct(new KeyValueEqualityComparer())
                    .ToDictionary(pair => pair.Key, pair => pair.Value);
```

msiarko replied:
Исправлено

KatzuoOgust commented on ReadingList.Domain/Services/Validation/OnNotExistExceptionData.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace ReadingList.Domain.Services.Validation
+{
+    public class OnNotExistExceptionData
```
можно сделать чайлд класс:
OnNotExistExceptionData<T>
и убрать `typeof(PrivateBookListItemWm).Name,` при вызове конструктора ибо вытянется из typeof(T).Name

KatzuoOgust commented on ReadingList.Domain/Services/Validation/OnNotExistExceptionData.cs:
```diff
@@ -0,0 +1,35 @@
+
+        public string Params
+        {
+            get
```
учитывая что Params делает вызов рефлексии (тяжеловат в выполнении) лучше сделать его в виде метода
GetParams дабы подчеркнуть что он computed

KatzuoOgust commented on ReadingList.ReadModel/DbConnection/IReadDbConnection.cs:
```diff
@@ -0,0 +1,19 @@
+
+        Task<T> QueryFirstAsync<T>(string query, object param = null);
+
+        Task<TR> QuerySingleAsync<TP, TS, TR>(string query, Func<TP, TS, TR> map, object param = null);
```
TP, TS, TR - не самые описательные имена что есть что и должны быть полными, например TSource, TResult

KatzuoOgust commented on ReadingList.Tests/ExceptionToStatusCodeMapTests.cs:
```diff
@@ -0,0 +1,54 @@
+        }
+
+        [Fact]
+        public void MapNullReferenceExceptionType()
```
названия методов в тестах могут использовать нейминг гайдлайны отличные от стандартных
ибо имена могут быть многословны т.к. должны описывать кейс.
вот ряд примеров:
`<Method>_<ExpectedResult>_<Condition>`
`Map_ReturnsAbc_WhenXyzPassed`
`Map_ReturnsAbc123_WhenXyzAns0Passed`
`map_returns_smth_when_smth`

KatzuoOgust commented on ReadingList.Domain/Infrastructure/Extensions/UpdateEntityExtension.cs:
```diff
@@ -0,0 +1,45 @@
+            if(source == null)
+                throw new ArgumentNullException(nameof(source));
+            
+            var entityProperties = entity.GetType().GetProperties(BindingFlags.Public | BindingFlags.Instance);
```
учитывая что Enitiy не так и много то это дело можно и закешить/заюзать скомпиленные экспрешоны 
ибо рефлекшон не так быстр (пока что)

KatzuoOgust commented on ReadingList.Domain/CommandHandlers/PrivateList/UpdatePrivateListItemCommandHandler.cs:
```diff
@@ -0,0 +1,48 @@
+            var item = await _dbContext.PrivateBookListItems.FirstOrDefaultAsync(i =>
+                i.BookList.Owner.Login == command.UserLogin && i.Id == command.ItemId);
+
+            EntitiesValidator.Validate(item,
```
поидее можно было бы просто юзать throw new ObjectNotExistException как и было
не совсем понимаю почему не делать формирование месседжа в самом эксепшоне а ключ передавать в виде объекта/словаря и сериализовать в JSON к примеру

[ArgumentException для размышлений](https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/ArgumentException.cs)

KatzuoOgust commented on ReadingList.Client/src/index.tsx:
```diff
@@ -0,0 +1,20 @@
+
+ReactDOM.render(
+  <Provider store={store}>
+      <ConnectedRouter history={history}>
```
отступ великоват

msiarko committed:
Updates performed:
 - Updated names of test methods;
 - Renamed method for private book item update from "Put" to "UpdateListItem";
 - Moved ValidateModelState attribute to controller level;
 - Code refactoring of Client project;
 - Moved not excist message search functionality to ObjectNotExistException class;
 - Removed EntitiesValidator.

msiarko committed:
Updates performed:
 - Updated generic paramers names in IReadDbConnection and ReadDbConnection.

msiarko committed:
Updates performed:
 - Code refactoring.

msiarko committed:
Updates performed:
 - Created demo compiled version of Update method for entities.

msiarko committed:
Updates performed:
 - Implemented cheking of IgnoreUpdateAttribute in UpdateExpr.

msiarko committed:
Updates performed:
 - Fully implemented EntityUpdateService.

msiarko committed:
Updates performed:
 - Added source cheking in EntityUpdateService.

msiarko committed:
Updates performed:
- Added settings for correct launching of the project from VS Code;
- Moved MediatR dependencies and initialization from Api project to Domain project;
- Added DomainService which will be used by Api project through a composition;
- Marked Owner property of BookList class as non-updated.

msiarko committed:
Updates performed:
- Code refactoring;
- Fixed incrementing of wrong index in inner loop of update expression.

msiarko committed:
Updates performed:
- Moved calling of reflection methods out from expression.

msiarko committed:
Updates performed:
 - Moved helper method infos to private fields of EntityUpdateService class.

KatzuoOgust commented on ReadingList.Domain/Services/EntityUpdate/EntityUpdateService.cs:
```diff
@@ -0,0 +1,132 @@
+
+        private Action<TEntity, Dictionary<string, object>> GetUpdater<TEntity>() where TEntity : Entity
+        {
+            if (_cache.ContainsKey(typeof(TEntity)))
```
EntityUpdateService  - зареген как синглтон 
а доступ к кешу не thread-safe 
можно заюзать ConcurrentDictionary

KatzuoOgust commented on ReadingList.Domain/Services/EntityUpdate/EntityUpdateService.cs:
```diff
@@ -0,0 +1,132 @@
+                                new[] {value},
+                                Expression.Assign(value,
+                                    Expression.Call(_getValueOrDefaultMethodInfo, sourceParameter, Expression.Property(property, "Name"))),
+                                Expression.Call(property, _setValueMethodInfo, entityParameter, value))),
```
все равно юзаешь рефлексию в скомпиленном экспрешоне,
у Call сигнатура подсказывает: `Expression.Call(Expression instance, MethodInfo, method, ...)`
что ты вызываешь метод инстанса т.е. что-то вроде `property.SetValue(...)`

должно быть что-то вроде такого :

```cs
var propertyExpression = Expression.PropertyOrField(entityParameter, property.Name);
Expression.Assign(
  propertyExpression, 
  Expression.Call(_getValueOrDefaultMethodInfo, sourceParameter, Expression.Property(property, "Name"))
)
```

пересмотри создание апдейтера, кмк он должен выглядеть так:
ты указываешь TEntity а билдер достает все пропы
затем фильтрует и оставляет те которые моджно сетать
и на основе этого списка строит блок экспрешон с кучей Assign выражений
одно Assign выражение сетает конкретный проп если есть значение в словаре (а если его нет то мб сетает дефолт либо не меняет - это уже надо уточнять)
и уже построенный блок компилит

msiarko committed:
Updates performed:
- Optimized Update mthod of EntityUpdateService;
- Code refactoring.

KatzuoOgust commented on ReadingList.Domain/Services/EntityUpdate/EntityUpdateService.cs:
```diff
@@ -47,86 +40,52 @@ public EntityUpdateService()
-            if (_cache.ContainsKey(typeof(TEntity)))
+            var entityType = typeof(TEntity);
+
+            if (_cache.TryGetValue(entityType, out var updater))
```
лучше использовать `GetOrAdd` https://msdn.microsoft.com/en-us/library/ee378677(v=vs.110).aspx
вместо  нескольких `TryGetValue`  и `TryAdd`
вопрос атомарности с точки зрения работы из разных потоков

KatzuoOgust commented on ReadingList.Domain/Services/EntityUpdate/EntityUpdateService.cs:
```diff
@@ -47,86 +40,52 @@ public EntityUpdateService()
-            var property = Expression.Parameter(typeof(PropertyInfo), "property");
-            var value = Expression.Parameter(typeof(object), "value");
+            var propertyInfos = entityType.GetProperties(BindingFlags.Instance | BindingFlags.Public).Where(p =>
+                p.CustomAttributes.All(a => a.AttributeType != typeof(IgnoreUpdateAttribute))).ToList();
```
можно проверить используя [экстеншон](https://msdn.microsoft.com/en-us/library/hh138307(v=vs.110).aspx): 
`p.GetCustomAttribute<IgnoreUpdateAttribute>(false) == null`

KatzuoOgust commented on ReadingList.Domain/Services/EntityUpdate/EntityUpdateService.cs:
```diff
@@ -47,86 +40,52 @@ public EntityUpdateService()
+            {
+                var propertyName = Expression.Constant(propertyInfo.Name);
+                var propertyExpression = Expression.PropertyOrField(entityParameter, propertyInfo.Name);
+                expressions.Add(Expression.IfThen(
```
можно использовать `yield return` вместо добавления в лист

msiarko committed:
Updates performed:
 - Code refactoring.

msiarko committed:
Updates performed:
- Changed reading time data type since TimeSpan type was projected by EF to time db data type which can't save more than 24 hours;
- Updated reading time displaying on UI to format "days | hours | minutes".

msiarko committed:
Updates performed:
- Started implementation of shared lists feature;
- Code refactoring.

msiarko committed:
Updates performed:
- Added GetUserSharedLists method in API;
- Code refactoring.

msiarko committed:
Updates performed:
- Added new methods in API for shared lists;
- Code refactoring.

msiarko committed:
Updates performed:
- Code refactoring.

msiarko committed:
Updates performed:
- Code refactoring;
- Applied convention of model names;
- Separated connection string to Read and Write connection strings in appsettings.json;
- Refactored exception messages generation.

msiarko committed:
Updates performed:
- Code refactoring;
- Changed routes;
- Optimized exception message generation when object doesn't exist;
- Added ObjectNotExist exceptions tests.

msiarko committed:
Updates performed:
- Added custom route attributes;
- Added route builders;
- Added BookInfo class which contains base book info (title and author). Class was added for decreasing methods parameters count;
- Code refactoring.

msiarko committed:
Code refactoring

msiarko committed:
Updates performed:
- Code refactoring;
- Added UpdateSharedListCommand;
- Added UpdateSharedListItemCommand;
- Added tags and category to the shared list query data;
- Added helper table for shared list and tags.

msiarko committed:
Updates performed:
- Code refactoring;
- Added new methods to API project;
- Added Category property to the shared list item.

msiarko committed:
Updates performed:
- Code refactoring;
- Removed JsonFields from BookList entity;
- Renamed Category to Genre;
- Added genres seeding to database.

msiarko committed:
Updates performed:
- Code refactoring;
- Added moderators;
- Added book list access validation.

msiarko committed:
Updates performed:
- Code refactoring;
- Added book statuses controller.

msiarko committed:
Updates performed:
- Code refactoring.

msiarko committed:
Updates performed:
- Code refactoring;
- Added expression filters;
- Added Genre property to the Private List Item;
- Added AccessDeniedException to ExceptionToStatusCode map.

msiarko committed:
Updates performed:
- Code refactoring.

msiarko committed:
Updates performed:
Server-side:
- Code refactoring;
- Fixed ToSelectListItems method for enums;
- Added tests for book items exception message generating.
Client-side:
- Code refactoring;
- Changed the way of getting book list statuses;
- Started work onshared book lists components.

msiarko committed:
Added missing changes

msiarko committed:
Updates prformed:
Client-side:
- Implemented Shared Lists page;
- Code refactoring.
Server-side:
- Added FindSharedBookLists method to the API (search hasn't been implemented yet).

msiarko committed:
Client:
- Code refactoring;
- Changed favicon.ico.

msiarko committed:
API:
- Code refactoring;
- Update/Add command now returns updated/added DTO;
- IReadConnection/ReadConnection renamed to IDbReader/DbReader.
Client:
- Code refactoring;
- Style changes.

msiarko committed:
API:
- Fixed issue with unmapped field "GenreId".

msiarko committed:
API:
- Merged migrations;
- Replaced MS SQL Localdb engine with SqLite (to make a cross platform development available);
- Code refactoring.

msiarko committed:
API:
- Changed some shared lists API methods signatures;
- Code refactoring.
Client:
- Updated routing.

msiarko committed:
Client:
- Code refactoring;
- Separated loading actions from api requests.

msiarko committed:
Client:
- Code refactoring;
- Added SharedBookList component.

msiarko committed:
Client:
- Code refactoring;
- Added count of books for each book list at Shared Lists grid.

msiarko committed:
API:
- Fixed issue with count of book in own book lists query;
- Fixed displaing of previuos query in search input.

msiarko committed:
Added code for launching the Client project when the API project starts

msiarko committed:
Merge branch 'develop' into shared

msiarko committed:
Fixed launching of client project on different platforms

msiarko committed:
Merge branch 'develop' into shared

msiarko committed:
Removed shutdown timeout

msiarko committed:
Merge branch 'develop' into shared

msiarko committed:
Client:
- Removed unnecessary API call in Account container;
- Updated approach of showing Spinner comp[onent.

msiarko committed:
API:
- Removed launching client project from code since it works only for Windows platform.

msiarko committed:
Merge branch 'develop' into shared

msiarko committed:
API:
- Removed unused field.

msiarko committed:
Merge branch 'develop' into shared

msiarko committed:
API:
- Backed deleted host run method.

msiarko committed:
Merge branch 'develop' into shared

msiarko committed:
Client:
- Added creation of shared list geature;
- Code refactoring.

msiarko committed:
Added universal form for adding elements

- Replaced ItemForm component with AddForm component;
- Removed ItemForm component.

msiarko committed:
Merge branch 'private' into shared

msiarko committed:
Replaced SharedListForm component with AddFrom component

msiarko committed:
Code refactoring and fixing of some style issues

- Added redirection from login or register page in case if user is authenticated;
- Changed styles of SharedBookListLI component;
-Code refactoring.

msiarko committed:
API:
- Changed solution structure;
- Code refactoring.
Client:
- Code refactoring.

msiarko committed:
API:
- Commit with deleted projects.

msiarko committed:
API:
- Fixed wrong namespace and project settings of Write project

msiarko committed:
Code refactoring

msiarko committed:
Added forgotten file

msiarko committed:
Client: - UI refactoring; Added feature of adding items to shared lists.

msiarko committed:
API:
- Refactored ReadingTimeCalculator class;
- Implemented assigning of CaEdit field in SharedBookListDto class.
Client:
- Added validation if user is owner of the shared list.

msiarko committed:
API:
- Moved updating of entity to the Domain project, now each entity can update itself

msiarko committed:
API:
- Redesigned solution structure;
- Redesigned read and write functionality;
- Domain project is fully independent from specific DB and its access functionality;
- Changed entities relations;
- Database creation and migrating is independent from specific DB and its access functionality;
- Code refactoring.
Client:
- Code refactoring;
- Updated functions linked with entity relations according to domain model changes.

msiarko committed:
API:
- Fixed incorrect read queries;
- Fixed registration issue;
- Code refactoring.
Client:
- Code refactoring.

msiarko committed:
API:
- Added SaveRangeAsync method to the IDataStorage interface;
- Fixed access validation issue when shared item added;
- Implemented tags saving to DB;
- Code refactoring.
Client:
- Updated code according to API changes.

msiarko committed:
API:
- Updated GetIncludePaths method to return only child navigation properties;
- Code refactoring.
Client:
- Replaced SharedLists and SharedList components routes with private routes components.

msiarko committed:
API:
- Implemented books search;
- Returned old GetIncludePaths method implementation;
- Code refactoring.
Client:
- Updated list comonents to reflect new searching functionality.

msiarko committed:
API:
- Fixed issue with saving entities in update command handlers;
- Code refactoring.
Client:
- Code refactoring;
- Decreased timeout value from 1000ms to 500ms.

msiarko committed:
API:
- Implemented SharedBookLists searching (but it has an issue when searching by a tag);
- Updated SharedBookList select by id query.
Client:
- Code refactoring;
- Updated UI according to new features.

msiarko committed:
API:
- Fixed issue related to genres.json path;
- Fixed issue related to viewing #null tag for books without any tags.

msiarko committed:
API:
- Fixed the issue when searching SharedBookList by a tag (set of tags of the shared list contained only one tag);
- Updated SharedBookList find SQL query;
- Code refactoring.

msiarko committed:
API:
- Fixed wrong book count for shared list query;
- Fixed the issue with hiding search items list when focus is lost.

msiarko committed:
Client:
- Code refactoring.

msiarko committed:
- Applied one code style to source code;
- Code refactoring;
- Added .editorconfig file.

msiarko committed:
API:
- Updated framework versions to .NET Core 2.1;
- Updated dependencies to the latest minor versions;
- Removed unused fetch query.

msiarko committed:
API:
- Removed unused reference.

msiarko committed:
API:
- Changed name convention;
- Performed code refactoring;
- Moved read and write models to separate project.
Client:
- Removed unused dependencies from package.json file;
- Added notifications.

msiarko committed:
Client:
- Code refacoring.

msiarko committed:
API:
- Created Tags controller;
Client:
- Started implementation of multi-select component with ability to add new options;
- Fixed issue with API calling loop in some shared lists component;
- Refactored using of Spinner component.

msiarko committed:
Fixed Spinner vertical alignment.

msiarko committed:
Client:
- Updated AddForm to work with select elements;
- Code refactoring;
- Finished implementation of multi-select component with ability to add new options.

msiarko committed:
API:
- Updated nuget packages;
- Updated target framework to .NET Core 2.2.

msiarko committed:
API:
- Removed unused query.
Client:
- Update styling of some components;
- Added context menu to Shared Lists Grid items;
- Fixed withContextMenu HOC.

msiarko committed:
Client:
- Implemented editing of the shared list data (name and tags).

msiarko committed:
API:
- Removed unused query;
- Added shared list name specification.
Client:
- Update styling of some components;
- Changed the way of adding books to lists;
- Added Genre to book view.

msiarko committed:
API:
- Added sharing of the private list.
Client:
- Added sharing action button;
- Added sharing form;
- Code refactoring.

msiarko committed:
Committed forgotten files

msiarko committed:
- Implemented moderators assigning for shared lists;
- Separated permissions related to editing and moderating shared list;
Client:
- Updated some UI styles;
- Code refactoring.

msiarko committed:
Committed forgotten files

msiarko committed:
API:
- Added reading of the book genre in SQL queries.
Client:
- Updated UI styling of some conponents;
- Code refactoring.

msiarko committed:
API:
- Added functionality for adding book to list;
- Added posibility to get all lists which can be moderated by user.

msiarko committed:
API:
- Implemented chunked request of the shared lists.
Client:
- Added pagination on SharedLists page.

msiarko committed:
Fixed wrong dependency in package json

msiarko committed:
API:
- Added posibility to share book with other lists moderated by user (known issue: after sharing Dapper can't read the list correctly);
Client:
- Implemented form for sharing book from one list to another;
- Fixed bug in MultiSelect component.

msiarko committed:
Merge branch 'shared' of https://github.com/MikhailSiarko/ReadingList into shared

msiarko committed:
API:
- Updated books query.
Client:
- Added pagination to the books form.

msiarko committed:
API:
- LoginUser is a command now;
- Updated GetUserId extension method.

msiarko committed:
Fixed an exception after submitting of the user data at the registration page

msiarko committed:
Updated Shared book list item mapping test

msiarko committed:
Fixed Grid component and updated Books and Shared Lists read queries.

msiarko committed:
Added removing of the shared list.

msiarko committed:
Updated types and responses some of the API methods in order to better match the semantics and fixed issue with list items sharing.

msiarko committed:
Code refactoring and bug fixing:
- Added OwnerLogin selecting for lists read queries;
- Fixed issue with wrong page when searching;
- Domain and Read project now are responsible for registration of the Command/Query validators.

msiarko committed:
Fixed wrong files encoding and updated some tests

msiarko committed:
Fixed exception throwing when lists fetching is failed

msiarko committed:
Added react-pose animation library and updated TypeScript version:
- Added animation to the Spinner component.

msiarko committed:
Removed unused globals from package.json

msiarko committed:
Merge branch 'develop' into animation

msiarko committed:
Updated UI styles

msiarko committed:
Merge branch 'develop' into animation

msiarko committed:
Started working on rewriting of the styles with Sass

msiarko committed:
Changed GetUserLists route to 'api/lists/shared/my'

msiarko committed:
CSS completely replaced with Sass:
- Code refactoring;
- Fixed issue with wrong argument name in PrivateListController.cs file.

msiarko committed:
Merge branch 'styling' into develop

msiarko committed:
Code refactoring:
- All sizes are based on view height;
- Updated styling.

msiarko committed:
Started implemetion of the side-effect management with react-saga library:
Implemented for:
- Notifications;
- Book fetching;
- Modetated lists fetching;
- Private list fetching and updates;
- Authentication.

msiarko committed:
Code refactoring of the saga common functions

msiarko committed:
Side-effect management is fully implemented:
- Added management for shared list;
- Code refactoring;
- Removed unused template generation.

msiarko committed:
Code refactoring

msiarko committed:
Code refactored and added clearing of the book state after add book form closing

msiarko committed:
Deleted unused util methods

msiarko committed:
Merged 'styling' to 'develop' with conflits in Account.tsx

msiarko committed:
Updated API service and added tag state clearing action

msiarko committed:
Updated webpack production configuration

msiarko committed:
Added routes configuration

msiarko committed:
Changes:
- Added classnames library;
- Removed unused util methods;
- Code refactoring.

msiarko committed:
Code refactoring and bug fixes

msiarko committed:
Code refactoring and components re-organizing

msiarko committed:
Stylesheet refactoring
